tests/kafkatest/services/verifiable_consumer.py

```diff
@@ -92,55 +174,40 @@ def _worker(self, idx, node):
-                        self._handle_records_consumed(node, event)
-                    elif name == "partitions_revoked":
-                        self._handle_partitions_revoked(node, event)
-                    elif name == "partitions_assigned":
-                        self._handle_partitions_assigned(node, event)
-
-    def _handle_shutdown_complete(self, node):
-        if node in self.joined:
-            self.joined.remove(node)
-
-    def _handle_offsets_committed(self, node, event):
-        if event["success"]:
-            for offset_commit in event["offsets"]:
-                topic = offset_commit["topic"]
-                partition = offset_commit["partition"]
-                tp = TopicPartition(topic, partition)
-                self.committed_offsets[tp] = offset_commit["offset"]
-
-    def _handle_records_consumed(self, node, event):
-        for topic_partition in event["partitions"]:
-            topic = topic_partition["topic"]
-            partition = topic_partition["partition"]
-            tp = TopicPartition(topic, partition)
-            self.consumed_positions[tp] = topic_partition["maxOffset"] + 1
-        self.total_records += event["count"]
-
-    def _handle_partitions_revoked(self, node, event):
-        self.revoked_count += 1
-        self.assignment[node] = []
-        if node in self.joined:
-            self.joined.remove(node)
-
-    def _handle_partitions_assigned(self, node, event):
-        self.assigned_count += 1
-        self.joined.add(node)
-        assignment =[]
-        for topic_partition in event["partitions"]:
-            topic = topic_partition["topic"]
-            partition = topic_partition["partition"]
-            assignment.append(TopicPartition(topic, partition))
-        self.assignment[node] = assignment
+            with self.lock:
+                name = event["name"]
+                if name == "shutdown_complete":
+                    handler.handle_shutdown_complete()
+                if name == "offsets_committed":
+                    handler.handle_offsets_committed(event)
+                    self._update_global_committed(event)
+                elif name == "records_consumed":
+                    handler.handle_records_consumed(event)
+                    self._update_global_position(event)
+                elif name == "partitions_revoked":
+                    handler.handle_partitions_revoked(event)
+                elif name == "partitions_assigned":
+                    handler.handle_partitions_assigned(event)
+
+    def _update_global_position(self, consumed_event):
+        for consumed_partition in consumed_event["partitions"]:
+            tp = TopicPartition(consumed_partition["topic"], consumed_partition["partition"])
+            if tp in self.global_committed:
+                # verify that the position never gets behind the current commit.
+                # it would be nice to also verify that the position increases monotonically
```

Comment:
Even if you can't guarantee this in case of hard client failures, perhaps you could still have a check and log it (warning level?)